src/jltypes.c

```diff
@@ -1655,11 +1578,68 @@ jl_value_t *jl_type_intersection_matching(jl_value_t *a, jl_value_t *b,
     return *pti;
 }
 
-// --- type instantiation and cache ---
+JL_DLLEXPORT jl_value_t *jl_type_intersection(jl_value_t *a, jl_value_t *b)
+{
+    jl_svec_t *env = jl_emptysvec;
+    JL_GC_PUSH1(&env);
+    jl_value_t *ti = jl_type_intersection_matching(a, b, &env, jl_emptysvec);
+    JL_GC_POP();
+    return ti;
+}
+#endif
+
+jl_value_t *jl_type_intersection_matching(jl_value_t *a, jl_value_t *b,
+                                          jl_svec_t **penv, jl_svec_t *tvars)
+{
+    int sza = jl_subtype_env_size(a);
+    int szb = jl_subtype_env_size(b);
+    int sz = 0;
+    jl_value_t **env = (jl_value_t**)alloca((sza>szb?sza:szb)*sizeof(jl_value_t*));
+    jl_value_t *ans = NULL;
+    if (jl_subtype_env(a, b, env, szb)) {
+        ans = a;
+        sz = szb;
+    }
+    else if (jl_subtype_env(b, a, env, sza)) {
+        ans = b;
+        sz = sza;
+    }
+    else if (jl_is_leaf_type(a) || jl_is_leaf_type(b)) {
+        return jl_bottom_type;
+    }
+    if (ans == NULL) {
+        ans = b;
+        sz = szb;
+        int i=0;
+        while (jl_is_unionall(b)) {
+            env[i++] = (jl_value_t*)((jl_unionall_t*)b)->var;
+            b = ((jl_unionall_t*)b)->body;
+        }
+    }
+    jl_svec_t *e = jl_alloc_svec(sz);
+    *penv = e;
+    int i;
+    for(i=0; i < sz; i++)
+        jl_svecset(e, i, env[i]);
+    return ans;
+}
+
+JL_DLLEXPORT jl_value_t *jl_type_intersection(jl_value_t *a, jl_value_t *b)
+{
+    if (jl_subtype(a, b))
+        return a;
+    if (jl_subtype(b, a))
+        return b;
+    if (jl_is_leaf_type(a) || jl_is_leaf_type(b))
+        return jl_bottom_type;
+    return b;
```

Comment:
I love the simplicity, but I don't understand this at all. What happens if this gets called with `a=Tuple{Int,Real}` and `b=Tuple{Real,Int}`?